### PR TITLE
Add better support for default signatures in class definitions

### LINKED ISSUE #439: Show default signatures for class members
Especially _minimal complete definition: Nothing_, when there are generics involved is misleading.

### LINKED ISSUE #567: Haddock regression in GHC 8.0.1: redundant documentation for default signatures
With this file:

```haskell
{-# LANGUAGE DefaultSignatures #-}
-- | Documentation for Bug
module Bug where

-- | Documentation for Foo
class Foo a where
  -- | Documentation for bar
  bar :: a -> String
  -- | Documentation for bar's default implementation
  default bar :: Show a => a -> String
  bar = show
```

If `haddock-2.16.1` (shipped with GHC 7.10.3) on it, then it renders like so:

haddock-old

But if I run `haddock-2.17.2` (shipped with GHC 8.0.1) on it, then it renders like so:

haddock-new

I certainly wouldn't have expected the current behavior. The ideal would be one of the proposals in https://github.com/haskell/haddock/issues/439#issuecomment-135090049 (or, failing that, reverting back to the old behavior in `haddock-2.17.2`).

### PR DESCRIPTION
**Pull request by RyanGlScott**
Take two (see @mac-adder's #692 for take one). Fixes #439 and fixes #567.

This almost works without a hitch except for one glitch: Haddock has trouble finding documentation for a default signature. For instance, this file:

```haskell
{-# LANGUAGE DefaultSignatures #-}

module DefaultSignatures where

-- | Documentation for Foo.
class Foo a where
  -- | Documentation for bar and baz.
  bar, baz :: a -> String

  -- | Documentation for the default signature of bar.
  default bar :: Show a => a -> String
  bar = show

  -- | Documentation for baz'.
  baz' :: String -> a

  -- | Documentation for the default signature of baz'.
  default baz' :: Read a => String -> a
  baz' = read
```

Gets rendered as:

haddock

Notice that the docstrings for the default signatures have been dropped. I've tried to determine why this is happening—it appears that Haddock drops all docstrings that aren't found in an `AvailInfo`, but default signatures don't correspond to `AvailInfo`s. @alexbiehl, do you know how one might get that to work?

### CODE REVIEW

**Comment by RyanGlScott** [comment]
As far as I could see, yes.

File: haddock-api/src/Haddock/GhcUtils.hs [hunk 1]
```diff
@@ -112,6 +118,14 @@ sigNameNoLoc (InlineSig    n _)        = [unLoc n]
 sigNameNoLoc (FixSig (FixitySig ns _)) = map unLoc ns
 sigNameNoLoc _                         = []
 
+
+uniquifyName :: SetName name => name -> name
+uniquifyName = liftA2 setName (updateName . getName) id
+  where
+    updateName = liftA2 setNameUnique id $
+        liftA2 deriveUnique id ((+1) . getKey) . nameUnique
```
Comment (harpocrates) [inline haddock-api/src/Haddock/GhcUtils.hs]: This is a bit suspicious. That said, I think the smell is coming more from the fact that `deriveUnique` exists... and I don't see another easy workaround.

**Comment by harpocrates** [comment]
I've solved the issue around `AvailInfo` by manually looking up the docs for "uniquified" default methods names and putting them in the right `ExportItem`. The output of the test case now looks correct.

screen shot 2018-07-03 at 8 20 46 pm

However, there is something else we had not considered: the `Binary Name` instance does not distinguish between `name` and `uniquifyName name`, so the interface file ends up containing docs for default method XOR docs for method. It looks like this clever hack is falling apart all too soon. We could probably work around this by patching the `Binary InstalledInterface` instance (ex. splitting `docMap` into `docMap` and `docMapDefaults`), but I suspect this is going to be a recurring source of confusion/bugs.

Thoughts?

**Comment by sjakobi** [comment]
> However, there is something else we had not considered: the Binary Name instance does not distinguish between name and uniquifyName name, so the interface file ends up containing docs for default method XOR docs for method. It looks like this clever hack is falling apart all too soon. We could probably work around this by patching the Binary InstalledInterface instance (ex. splitting docMap into docMap and docMapDefaults), but I suspect this is going to be a recurring source of confusion/bugs.

If we split `docMap`, we'd have to split `argMap` too…

Maybe we can instead use a different key type for these maps.

Something like

```
data DocName = Name Name | DefaultSignatureName Name
```

maybe?

**Comment by sjakobi** [comment]
This should fix #30 too.

**Comment by RyanGlScott** [comment]
Was this subsumed by #938? If so, should this be closed?

**Comment by harpocrates** [comment]
There is still no support for documentation on default methods. #938 just ported over the HTML backend changes from this PR (basically parity with your original screenshot from here: https://github.com/haskell/haddock/pull/822#issue-187009263).

However, I'm now thoroughly convinced that the path forward to getting docs on default methods is going to involve their default method names (aka. `isDefaultMethodOcc . occName` returns `True`) and not [some brittle hack involving "uniquified" names](https://github.com/haskell/haddock/pull/822#discussion_r199855357).

I don't think there is much more to gain from keeping this PR open (please re-open if you disagree).

**Pull request closed without merging**